### ISSUE TITLE
[Request] Add option to prevent disk wipe (cancel install) when only one disk is present

### ISSUE THREAD

**Issue by RVRX** [cschneegans/unattend-generator#150]
Currently with the "_Partitioning and formatting_" option "_Let Windows Setup wipe, partition and format your hard drive (more specifically, disk 0) using these settings_" if you:
* are running the autoinstall via an external drive (usb) plugged into your machine
* and if Windows cannot detect the internal disk of your computer (e.g., missing driver, no disk)

Then
* The installation media will be disk 0, and as such, it will get wiped
* [displays the "diskpart encountered an error" msg & fails to continue]

Suggested solution:
* have an option that would cancel the install if only 1 disk (disk 0) is present.

**Comment by cschneegans** [comment]
I agree that the autounattend.xml process should not wipe an USB drive.

However, the suggested solution would not work since VM installations typically use optical drives, and these do not show up in diskpart. Therefore, having only disk 0 is the expected condition for VMs.

At the moment, I do not see any viable solution for this.

**Comment by R0GGER** [comment]
PC / Laptops:    
This problem may also occur because Intel Rapid Storage Technology (RST) is enabled in the BIOS. Disable RST and enable AHCI in the BIOS under storage settings… Autounattend will likely work after that.

**Comment by skey-scottlogic** [comment]
Here's what I'm doing to prevent the problem @R0GGER mentioned:

Write two separate diskpart scripts - a check script that tries to select all of the disks you expect to find, then the one that does the actual partitioning.  Run the check script and conditionally the partition script, using "&&".  Display an error, pause and exit if either fails.

Perhaps the xml generator could have a field for expected number of disks, and generate the check script based on that value - 1 for a VM installing from DVD, 2 for a laptop installing from USB, etc?

`<RunSynchronous>
	<RunSynchronousCommand wcm:action="add">
		<Order>1</Order>
		<Path>cmd.exe /c "&gt;&gt;"X:\diskpart-check.txt" (echo SELECT DISK=0&amp;echo SELECT DISK=1)"</Path>
	</RunSynchronousCommand>
	<RunSynchronousCommand wcm:action="add">
		<Order>2</Order>
		<Path>cmd.exe /c "&gt;&gt;"X:\diskpart.txt" (echo SELECT DISK=0&amp;echo CLEAN&amp;echo CONVERT GPT&amp;echo CREATE PARTITION EFI SIZE=300&amp;echo FORMAT QUICK FS=FAT32 LABEL="System"&amp;echo CREATE PARTITION MSR SIZE=16)"</Path>
	</RunSynchronousCommand>
	<RunSynchronousCommand wcm:action="add">
		<Order>3</Order>
		<Path>cmd.exe /c "&gt;&gt;"X:\diskpart.txt" (echo CREATE PARTITION PRIMARY&amp;echo SHRINK MINIMUM=1000&amp;echo FORMAT QUICK FS=NTFS LABEL="Windows"&amp;echo CREATE PARTITION PRIMARY&amp;echo FORMAT QUICK FS=NTFS LABEL="Recovery")"</Path>
	</RunSynchronousCommand>
	<RunSynchronousCommand wcm:action="add">
		<Order>4</Order>
		<Path>cmd.exe /c "&gt;&gt;"X:\diskpart.txt" (echo SET ID="de94bba4-06d1-4d40-a16a-bfd50179d6ac"&amp;echo GPT ATTRIBUTES=0x8000000000000001)"</Path>
	</RunSynchronousCommand>
	<RunSynchronousCommand wcm:action="add">
		<Order>5</Order>
		<Path>cmd.exe /c "(diskpart.exe /s "X:\diskpart-check.txt" &gt;&gt;"X:\diskpart.log" &amp;&amp; diskpart.exe /s "X:\diskpart.txt" &gt;&gt;"X:\diskpart.log") || ( type "X:\diskpart.log" &amp; echo diskpart encountered an error. You need to start again. Check that AHCI is selected for storage. &amp; pause &amp; exit /b 1 )"</Path>
	</RunSynchronousCommand>
</RunSynchronous>`

**Comment by cschneegans** [comment]
Using SELECT DISK=… to assert that the disks are present as expected is an interesting approach.

However, I don't see the benefit of having two diskpart script files, diskpart-check.txt and diskpart.txt. If you add

```
SELECT DISK=0
SELECT DISK=1
```

at the beginning of diskpart.txt, the effect will be the same: diskpart will fail and quit after the first error.

**Comment by cschneegans** [comment]
I have just added (via https://github.com/cschneegans/unattend-generator/commit/6c42986bedad2bebbd5561b6c23a12f50e25d0d8) a mechanism to provide a VBScript to check certain properties of your disks, typically from the [Win32_DiskDrive](https://learn.microsoft.com/en-us/windows/win32/cimwin32prov/win32-diskdrive) class. If that script returns with a non-zero error code, Windows Setup halts, and no changes are made to the system.

![Image](https://github.com/user-attachments/assets/0c0aa8b9-0f05-442b-8c0c-8589bcc939f1)

Note that this mechanism only ensures that a given disk was assigned the expected index. You still have to use that index in your diskpart script or the _Choose partition to install Windows to after script has run_ section.
